File: C++/22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */
#include <vector>
#include <string>
#include <iostream>
using namespace std;
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        string temp = "";
        generateParenthesisRec(n, n, res, temp);
        return res;
    }
private:
    // choose the right signal to loop
    void generateParenthesisRec(int left, int right, vector<string>& res, string& temp)
    {
        if(left < 0 || right < 0)
            return;
        // diff means the number of '(' minus the number of ')'
        if(left == 0 && right == 0)
        {
            res.push_back(temp);
            return;
        }
        if(left > 0)
        {
            temp.append("(");
            generateParenthesisRec(left - 1,right, res, temp);
            temp.pop_back();
        }
        if(right > 0 && right > left)
        {
            temp.append(")");
            generateParenthesisRec(left, right - 1, res, temp);
            temp.pop_back();
        }
        
    }
};

// int main()
// {
//     Solution s;
//     vector<string> res = s.generateParenthesis(3);
    
//     return 0;
// }

File: Java/22.generate-parentheses.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=22 lang=java
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution {
    public List<String> generateParenthesis(int n) {
        // time complexity: O(2 ^ (2n))
        // space complexity: O(2n * 4 ^ n)
        List<String> res = new ArrayList<>();
        this.dfs(n, n, new StringBuilder(), res);
        return res;
    }

    private void dfs(int left, int right, StringBuilder sb, List<String> res) {
        if (left == 0 && right == 0) {
            res.add(sb.toString());
            return;
        }

        // check if we can add left parenthesis
        if (left > 0) {
            sb.append('(');
            dfs(left - 1, right, sb, res);
            sb.deleteCharAt(sb.length() - 1);
        }

        // check if we can add right parenthesis
        if (left < right) {
            sb.append(')');
            dfs(left, right - 1, sb, res);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}
// @lc code=end

